### PR TITLE
WW-348 add tracking for create-button flow for VE

### PR DESCRIPTION
**Pull request by cormorando**

### CODE REVIEW

File: extensions/VisualEditor/wikia/modules/ve/init/ve.init.wikia.ViewPageTarget.init.js [hunk 1]
```diff
@@ -229,7 +229,7 @@
 
 			if ( history.pushState && uri.query.veaction !== 'edit' ) {
 				if (!pageExists) {
-					veEditUri.extend( { flow: 'create-page-create-button' } );
+					mw.hook('ve.afterVEInit').fire(veEditUri);
 				}
 				// Replace the current state with one that is tagged as ours, to prevent the
 				// back button from breaking when used to exit VE. FIXME: there should be a better
```
Comment (WikiaTech) [inline extensions/VisualEditor/wikia/modules/ve/init/ve.init.wikia.ViewPageTarget.init.js:235]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AS1134)

File: extensions/wikia/FlowTracking/FlowTracking.hooks.php [hunk 2]
```diff
@@ -85,6 +89,31 @@ public static function getParamsFromUrlQuery( $url ) {
 		return $queryParams;
 	}
 
+	/**
+	 * @param PageHeaderController $header
+	 * @param $actions
+	 * @return bool
+	 */
+	public static function onBeforePrepareActionButtons( $header, &$actions ) {
```
Comment (WikiaTech) [inline extensions/wikia/FlowTracking/FlowTracking.hooks.php:97]: MAJOR Remove the unused function parameter "$header". [rule](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS1172)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 38 issues
* MAJOR 25 major
* MINOR 6 minor
* INFO 7 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [ve.init.wikia.ViewPageTarget.init.js#L57](https://github.com/Wikia/app/blob/24d1ade49f143fdb5c4d05cb3343b05432e5595c/extensions/VisualEditor/wikia/modules/ve/init/ve.init.wikia.ViewPageTarget.init.js#L57): Make this function anonymous by removing its name: 'function() {...}'. [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3ANamedFunctionExpression)
1. MAJOR [ve.init.wikia.ViewPageTarget.init.js#L126](https://github.com/Wikia/app/blob/24d1ade49f143fdb5c4d05cb3343b05432e5595c/extensions/VisualEditor/wikia/modules/ve/init/ve.init.wikia.ViewPageTarget.init.js#L126): Remove this useless assignment to local variable "tabMessages" [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AS1854)
1. MAJOR [ve.init.wikia.ViewPageTarget.init.js#L144](https://github.com/Wikia/app/blob/24d1ade49f143fdb5c4d05cb3343b05432e5595c/extensions/VisualEditor/wikia/modules/ve/init/ve.init.wikia.ViewPageTarget.init.js#L144): Reduce the number of conditional operators (12) used in the expression (maximum allowed 3). [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AS1067)
1. MAJOR [ve.init.wikia.ViewPageTarget.init.js#L218](https://github.com/Wikia/app/blob/24d1ade49f143fdb5c4d05cb3343b05432e5595c/extensions/VisualEditor/wikia/modules/ve/init/ve.init.wikia.ViewPageTarget.init.js#L218): Function has a complexity of 12 which is greater than 10 authorized. [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AFunctionComplexity)
1. MAJOR [ve.init.wikia.ViewPageTarget.init.js#L223](https://github.com/Wikia/app/blob/24d1ade49f143fdb5c4d05cb3343b05432e5595c/extensions/VisualEditor/wikia/modules/ve/init/ve.init.wikia.ViewPageTarget.init.js#L223): Reduce the number of conditional operators (5) used in the expression (maximum allowed 3). [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AS1067)
1. MAJOR [ve.init.wikia.ViewPageTarget.init.js#L268](https://github.com/Wikia/app/blob/24d1ade49f143fdb5c4d05cb3343b05432e5595c/extensions/VisualEditor/wikia/modules/ve/init/ve.init.wikia.ViewPageTarget.init.js#L268): Function has a complexity of 11 which is greater than 10 authorized. [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AFunctionComplexity)
1. MAJOR [ve.init.wikia.ViewPageTarget.init.js#L269](https://github.com/Wikia/app/blob/24d1ade49f143fdb5c4d05cb3343b05432e5595c/extensions/VisualEditor/wikia/modules/ve/init/ve.init.wikia.ViewPageTarget.init.js#L269): Reduce the number of conditional operators (5) used in the expression (maximum allowed 3). [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AS1067)
1. MAJOR [ve.init.wikia.ViewPageTarget.init.js#L278](https://github.com/Wikia/app/blob/24d1ade49f143fdb5c4d05cb3343b05432e5595c/extensions/VisualEditor/wikia/modules/ve/init/ve.init.wikia.ViewPageTarget.init.js#L278): Take the required action to fix the issue indicated by this comment. [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AS1134)
1. MAJOR [ve.init.wikia.ViewPageTarget.init.js#L325](https://github.com/Wikia/app/blob/24d1ade49f143fdb5c4d05cb3343b05432e5595c/extensions/VisualEditor/wikia/modules/ve/init/ve.init.wikia.ViewPageTarget.init.js#L325): Reduce the number of conditional operators (4) used in the expression (maximum allowed 3). [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AS1067)
1. MAJOR [ve.init.wikia.ViewPageTarget.init.js#L400](https://github.com/Wikia/app/blob/24d1ade49f143fdb5c4d05cb3343b05432e5595c/extensions/VisualEditor/wikia/modules/ve/init/ve.init.wikia.ViewPageTarget.init.js#L400): Make this function anonymous by removing its name: 'function() {...}'. [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3ANamedFunctionExpression)

File: extensions/VisualEditor/wikia/modules/ve/init/ve.init.wikia.ViewPageTarget.init.js [hunk 3]
```diff
@@ -229,7 +229,7 @@
 
 			if ( history.pushState && uri.query.veaction !== 'edit' ) {
 				if (!pageExists) {
```
Comment (lukaszkonieczny) [inline extensions/VisualEditor/wikia/modules/ve/init/ve.init.wikia.ViewPageTarget.init.js:231]: What do you think about moving this if statement to the place where hook is invoked?

File: extensions/wikia/FlowTracking/scripts/createPageTracking.js [hunk 4]
```diff
@@ -41,7 +41,12 @@ define(
 			return namespaceId === -1 && title === 'CreatePage';
 		}
 
+		function extendUri(veEditUri) {
```
Comment (lukaszkonieczny) [inline extensions/wikia/FlowTracking/scripts/createPageTracking.js]: Do we need separate method for this?

File: extensions/VisualEditor/wikia/modules/ve/init/ve.init.wikia.ViewPageTarget.init.js (hunk 3 above)
Comment (cormorando) [reply extensions/VisualEditor/wikia/modules/ve/init/ve.init.wikia.ViewPageTarget.init.js:231]: I would leave it as is - we didn't want to change much of VE code in the first place.

**Comment by WikiaTech** [comment]
SonarQube analysis reported 37 issues
- MAJOR 25 major
- MINOR 6 minor
- INFO 6 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:
1. MAJOR [ve.init.wikia.ViewPageTarget.init.js#L57](https://github.com/Wikia/app/blob/98f14a7ef5f688e0b9b1e90672f38575408376be/extensions/VisualEditor/wikia/modules/ve/init/ve.init.wikia.ViewPageTarget.init.js#L57): Make this function anonymous by removing its name: 'function() {...}'. [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3ANamedFunctionExpression)
2. MAJOR [ve.init.wikia.ViewPageTarget.init.js#L126](https://github.com/Wikia/app/blob/98f14a7ef5f688e0b9b1e90672f38575408376be/extensions/VisualEditor/wikia/modules/ve/init/ve.init.wikia.ViewPageTarget.init.js#L126): Remove this useless assignment to local variable "tabMessages" [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AS1854)
3. MAJOR [ve.init.wikia.ViewPageTarget.init.js#L144](https://github.com/Wikia/app/blob/98f14a7ef5f688e0b9b1e90672f38575408376be/extensions/VisualEditor/wikia/modules/ve/init/ve.init.wikia.ViewPageTarget.init.js#L144): Reduce the number of conditional operators (12) used in the expression (maximum allowed 3). [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AS1067)
4. MAJOR [ve.init.wikia.ViewPageTarget.init.js#L218](https://github.com/Wikia/app/blob/98f14a7ef5f688e0b9b1e90672f38575408376be/extensions/VisualEditor/wikia/modules/ve/init/ve.init.wikia.ViewPageTarget.init.js#L218): Function has a complexity of 12 which is greater than 10 authorized. [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AFunctionComplexity)
5. MAJOR [ve.init.wikia.ViewPageTarget.init.js#L223](https://github.com/Wikia/app/blob/98f14a7ef5f688e0b9b1e90672f38575408376be/extensions/VisualEditor/wikia/modules/ve/init/ve.init.wikia.ViewPageTarget.init.js#L223): Reduce the number of conditional operators (5) used in the expression (maximum allowed 3). [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AS1067)
6. MAJOR [ve.init.wikia.ViewPageTarget.init.js#L268](https://github.com/Wikia/app/blob/98f14a7ef5f688e0b9b1e90672f38575408376be/extensions/VisualEditor/wikia/modules/ve/init/ve.init.wikia.ViewPageTarget.init.js#L268): Function has a complexity of 11 which is greater than 10 authorized. [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AFunctionComplexity)
7. MAJOR [ve.init.wikia.ViewPageTarget.init.js#L269](https://github.com/Wikia/app/blob/98f14a7ef5f688e0b9b1e90672f38575408376be/extensions/VisualEditor/wikia/modules/ve/init/ve.init.wikia.ViewPageTarget.init.js#L269): Reduce the number of conditional operators (5) used in the expression (maximum allowed 3). [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AS1067)
8. MAJOR [ve.init.wikia.ViewPageTarget.init.js#L278](https://github.com/Wikia/app/blob/98f14a7ef5f688e0b9b1e90672f38575408376be/extensions/VisualEditor/wikia/modules/ve/init/ve.init.wikia.ViewPageTarget.init.js#L278): Take the required action to fix the issue indicated by this comment. [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AS1134)
9. MAJOR [ve.init.wikia.ViewPageTarget.init.js#L325](https://github.com/Wikia/app/blob/98f14a7ef5f688e0b9b1e90672f38575408376be/extensions/VisualEditor/wikia/modules/ve/init/ve.init.wikia.ViewPageTarget.init.js#L325): Reduce the number of conditional operators (4) used in the expression (maximum allowed 3). [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AS1067)
10. MAJOR [ve.init.wikia.ViewPageTarget.init.js#L400](https://github.com/Wikia/app/blob/98f14a7ef5f688e0b9b1e90672f38575408376be/extensions/VisualEditor/wikia/modules/ve/init/ve.init.wikia.ViewPageTarget.init.js#L400): Make this function anonymous by removing its name: 'function() {...}'. [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3ANamedFunctionExpression)

File: extensions/wikia/FlowTracking/scripts/initCreatePageTracking.js [hunk 5]
```diff
@@ -6,11 +6,13 @@ require([
 	'jquery',
 	'wikia.window'
 ], function (flowTracking, flowTrackingCreatePage, QueryString, mw, $, window) {
+	var redLinkFlow = mw.config.get('wgNamespaceNumber') === -1 ?
+			window.wgFlowTrackingFlows.CREATE_PAGE_SPECIAL_REDLINK :
+			window.wgFlowTrackingFlows.CREATE_PAGE_ARTICLE_REDLINK,
+		createButtonFlow = window.wgFlowTrackingFlows.CREATE_PAGE_CREATE_BUTTON;
```
Comment (idradm) [inline extensions/wikia/FlowTracking/scripts/initCreatePageTracking.js:12]: This is a static value, why we assign it additionally to variable?

Comment (sebastianmarzjan) [reply extensions/wikia/FlowTracking/scripts/initCreatePageTracking.js:12]: Hmm, it's either leaning towards keeping flows as local vars (current approach) or using this static value in https://github.com/Wikia/app/pull/11589/files#diff-2e6e1784157fc2addcc6e0435179abedR36

I have no strong preference either way, both are readable

**Review by sebastianmarzjan — APPROVED** [review]

**Review by rybmat — APPROVED** [review]

**Comment by cormorando** [comment]
those changes were moved to https://github.com/Wikia/app/pull/11587

**Pull request closed without merging**